MRG+1: Use intersphinx

larsoner opened the pull request:
Closes #287.

Simplifies link resolution by using `intersphinx`.

~~I'm not sure the best way to test this. I was doing it locally by having a small `testbuild` directory with a `conf.py`, `index.rst`, `examples/plot_numpy_scipy.py`. I could turn this into a test if it seems useful. Would the best way be to directly include the files, copy them to a tempdir, call `Sphinx(...)`, and check the output links to ensure they contain links to numpy/scipy/matplotlib docs, as they should?~~

larsoner commented on doc/reference.rst:
```diff
@@ -9,6 +9,12 @@ every module.
+   :no-members:
+   :no-inherited-members:
+
+:py:mod:`sphinx_gallery`:
```
This allows other projects to link to the root module, which is nice, as:
```
:mod:`sphinx_gallery`
```

larsoner commented:
Updated [docs](https://197-25860190-gh.circle-artifacts.com/0/home/ubuntu/sphinx-gallery/rtd_html/advanced_configuration.html#link-to-documentation) and [example](https://197-25860190-gh.circle-artifacts.com/0/home/ubuntu/sphinx-gallery/rtd_html/auto_examples/plot_exp.html#sphx-glr-auto-examples-plot-exp-py).

GaelVaroquaux commented:
Trying to build scipy-lectures with this PR, I get the following traceback:
<pre>
# Sphinx version: 1.4.8
# Python version: 2.7.12+ (CPython)
# Docutils version: 0.12 release
# Jinja2 version: 2.8
# Last messages:
#   copying static files...
#   done
#   copying extra files...
#   done
#   dumping search index in English (code: en) ...
#   done
#   dumping object inventory...
#   done
#   build succeeded, 72 warnings.
#   embedding documentation hyperlinks in examples...
# Loaded extensions:
#   ipython_console_highlighting (unknown version) from /home/varoquau/dev/scipy-lecture-notes/sphinxext/ipython_console_highlighting.pyc
#   sphinx.ext.intersphinx (1.4.8) from /usr/lib/python2.7/dist-packages/sphinx/ext/intersphinx.pyc
#   sphinx.ext.doctest (1.4.8) from /usr/lib/python2.7/dist-packages/sphinx/ext/doctest.pyc
#   only_directives (unknown version) from /home/varoquau/dev/scipy-lecture-notes/sphinxext/only_directives.pyc
#   sphinx.ext.extlinks (1.4.8) from /usr/lib/python2.7/dist-packages/sphinx/ext/extlinks.pyc
#   sphinx_gallery.gen_gallery (unknown version) from /home/varoquau/dev/sphinx-gallery/sphinx_gallery/gen_gallery.pyc
#   sphinx.ext.autodoc (1.4.8) from /usr/lib/python2.7/dist-packages/sphinx/ext/autodoc.pyc
#   alabaster (0.7.9) from /home/varoquau/.local/lib/python2.7/site-packages/alabaster/__init__.pyc
#   sphinx.ext.imgmath (1.4.8) from /usr/lib/python2.7/dist-packages/sphinx/ext/imgmath.pyc
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/sphinx/cmdline.py", line 244, in main
    app.build(opts.force_all, filenames)
  File "/usr/lib/python2.7/dist-packages/sphinx/application.py", line 315, in build
    self.emit('build-finished', None)
  File "/usr/lib/python2.7/dist-packages/sphinx/application.py", line 551, in emit
    results.append(callback(self, *args))
  File "/home/varoquau/dev/sphinx-gallery/sphinx_gallery/docs_resolv.py", line 383, in embed_code_links
    _embed_code_links(app, gallery_conf, gallery_dir)
  File "/home/varoquau/dev/sphinx-gallery/sphinx_gallery/docs_resolv.py", line 266, in _embed_code_links
    src_gallery_dir)
  File "/home/varoquau/dev/sphinx-gallery/sphinx_gallery/docs_resolv.py", line 166, in __init__
    index = get_data(index_url, gallery_dir)
  File "/home/varoquau/dev/sphinx-gallery/sphinx_gallery/docs_resolv.py", line 73, in get_data
    data = _get_data(url)
  File "/home/varoquau/dev/sphinx-gallery/sphinx_gallery/docs_resolv.py", line 50, in _get_data
    data = StringIO(data)
TypeError: initial_value must be unicode or None, not str
</pre>

larsoner commented:
That is a bug already fixed by #293. I've rebased this PR to latest `master`, feel free to try again.

choldgraf commented on doc/conf.py:
```diff
@@ -305,6 +305,7 @@ def setup(app):
     'matplotlib': ('https://matplotlib.org/', None),
     'mayavi': ('http://docs.enthought.com/mayavi/mayavi', None),
     'sklearn': ('http://scikit-learn.org/stable', None),
+    'sphinx': ('http://www.sphinx-doc.org/en/stable', None),
```
could we add a comment above here that says something to the effect of "If you have intersphinx enabled, these lines are unnecessary"?

larsoner replied:
This is the standard `intersphinx_mapping` dict. I've already removed the unnecessary lines below

choldgraf commented:
LGTM with one comment

GaelVaroquaux commented:
On a clean build of scipy-lectures, using this PR, I get the following crasher:

<pre>
# Sphinx version: 1.4.8
# Python version: 2.7.12+ (CPython)
# Docutils version: 0.12 release
# Jinja2 version: 2.8
# Last messages:
#   embedding documentation hyperlinks for packages/statistics/auto_examples... [ 20%] plot_airfare.html
#   embedding documentation hyperlinks for packages/statistics/auto_examples... [ 30%] plot_iris_analysis.html
#   embedding documentation hyperlinks for packages/statistics/auto_examples... [ 40%] plot_paired_boxplots.html
#   embedding documentation hyperlinks for packages/statistics/auto_examples... [ 50%] plot_pandas.html
#   embedding documentation hyperlinks for packages/statistics/auto_examples... [ 60%] plot_regression.html
#   embedding documentation hyperlinks for packages/statistics/auto_examples... [ 70%] plot_regression_3d.html
#   embedding documentation hyperlinks for packages/statistics/auto_examples... [ 80%] plot_wage_data.html
#   embedding documentation hyperlinks for packages/statistics/auto_examples... [ 90%] plot_wage_education_gender.html
#   embedding documentation hyperlinks for packages/statistics/auto_examples... [100%] plot_brain_size.html
#   
# Loaded extensions:
#   ipython_console_highlighting (unknown version) from /home/varoquau/dev/scipy-lecture-notes/sphinxext/ipython_console_highlighting.pyc
#   sphinx.ext.intersphinx (1.4.8) from /usr/lib/python2.7/dist-packages/sphinx/ext/intersphinx.pyc
#   sphinx.ext.doctest (1.4.8) from /usr/lib/python2.7/dist-packages/sphinx/ext/doctest.pyc
#   only_directives (unknown version) from /home/varoquau/dev/scipy-lecture-notes/sphinxext/only_directives.pyc
#   sphinx.ext.extlinks (1.4.8) from /usr/lib/python2.7/dist-packages/sphinx/ext/extlinks.pyc
#   sphinx_gallery.gen_gallery (0.1.12) from /home/varoquau/dev/sphinx-gallery/sphinx_gallery/gen_gallery.pyc
#   sphinx.ext.autodoc (1.4.8) from /usr/lib/python2.7/dist-packages/sphinx/ext/autodoc.pyc
#   alabaster (0.7.9) from /home/varoquau/.local/lib/python2.7/site-packages/alabaster/__init__.pyc
#   sphinx.ext.imgmath (1.4.8) from /usr/lib/python2.7/dist-packages/sphinx/ext/imgmath.pyc
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/sphinx/cmdline.py", line 244, in main
    app.build(opts.force_all, filenames)
  File "/usr/lib/python2.7/dist-packages/sphinx/application.py", line 315, in build
    self.emit('build-finished', None)
  File "/usr/lib/python2.7/dist-packages/sphinx/application.py", line 551, in emit
    results.append(callback(self, *args))
  File "/home/varoquau/dev/sphinx-gallery/sphinx_gallery/docs_resolv.py", line 392, in embed_code_links
    _embed_code_links(app, gallery_conf, gallery_dir)
  File "/home/varoquau/dev/sphinx-gallery/sphinx_gallery/docs_resolv.py", line 264, in _embed_code_links
    url, src_gallery_dir)
  File "/home/varoquau/dev/sphinx-gallery/sphinx_gallery/docs_resolv.py", line 171, in __init__
    self._searchindex = js_index.loads(sindex)
  File "/usr/lib/python2.7/dist-packages/sphinx/search/__init__.py", line 146, in loads
    return jsdump.loads(data)
  File "/usr/lib/python2.7/dist-packages/sphinx/util/jsdump.py", line 187, in loads
    raise ValueError("read error at pos %d" % i)
ValueError: read error at pos 113500
</pre>

larsoner commented:
It looks unrelated to this PR. Does it pass on `master`?

GaelVaroquaux commented:
> It looks unrelated to this PR. Does it pass on master?

No, it doesn't pass. But it's basically the bug in #287, and I was hoping
that your PR was fixing it.

larsoner commented:
Get rid of the `reference_url` entry in `sphinx_gallery_conf`

larsoner commented:
(all of them, in this case)

larsoner commented:
... in other words, the standard `intersphinx_mapping`, which you appear to have defined properly already, should be sufficient. I still let `reference_url`s take precedence for reasons of backward compatibility.

GaelVaroquaux commented:
Indeed, of course; that works fine. Thank you!

GaelVaroquaux commented:
:+1: for merge

lesteve commented:
> I'm not sure the best way to test this. I was doing it locally by having a small testbuild directory with a conf.py, index.rst, examples/plot_numpy_scipy.py. I could turn this into a test if it seems useful. 

Being able to quickly set-up a toy gallery, run sphinx on it and check the output programatically would be really great. Historically for a few things, we have been relying on doing `make html` and checking manually. @Titan-C can correct me if I am wrong, maybe he has already added tests that were similar to what I describe.

Titan-C commented:
> Being able to quickly set-up a toy gallery, run sphinx on it and check
> the output programatically would be really great.

Indeed, I have tried to implement that. I reviewed how Sphinx tests
itself, it is not very nice. But it add up to create tiny documentation
sources and build them inside the test suite.

> maybe he has already added tests that were similar to what I describe.

In PR #225 I managed to implement this for the gallery to test at least
various options of the configuration file conf.py. It is in a usable
state to indeed create a testbuild directory and call Sphinx-build from
pytest.

Titan-C commented on doc/advanced_configuration.rst:
```diff
@@ -203,14 +203,17 @@ dictionary within your Sphinx ``conf.py`` file::
+If you do not use ``intersphinx``, then you should add entries that
+point to the directory containing ``searchindex.js``, such as
+``'matplotlib': 'https://matplotlib.org'``.
 
```
I will suggest that we do include a code example on this configuration style without intersphinx. That is just keep the old version of the configuration code example.

larsoner commented:
> Or much better integrate the test to our pytest tests.

This should be pretty straightforward with the Sphinx app. I'll give it a shot.

GaelVaroquaux commented:
> Removing the code sample is designed to help push people to really use intersphinx where possible.

I think that this is a good idea.

larsoner commented:
Added the `tinybuild` with `pytest` unit tests, [CircleCI rendering](https://204-25860190-gh.circle-artifacts.com/0/home/ubuntu/sphinx-gallery/tiny_html/index.html), and easy testing when running `make` within the `tinybuild` directory. I also added an auto-removing `_TempDir()` function ported from MNE (we've used it for years) to avoid filling up people's `/tmp`.

larsoner commented:
Travis is getting this error, anybody see a reason why? I don't get it locally:
```
==================================== ERRORS ====================================
______________________________ ERROR collecting  _______________________________
../../../miniconda2/envs/testenv/lib/python2.7/site-packages/py/_path/common.py:372: in visit
    for x in Visitor(fil, rec, ignore, bf, sort).gen(self):
../../../miniconda2/envs/testenv/lib/python2.7/site-packages/py/_path/common.py:421: in gen
    for p in self.gen(subdir):
../../../miniconda2/envs/testenv/lib/python2.7/site-packages/py/_path/common.py:421: in gen
    for p in self.gen(subdir):
../../../miniconda2/envs/testenv/lib/python2.7/site-packages/py/_path/common.py:421: in gen
    for p in self.gen(subdir):
../../../miniconda2/envs/testenv/lib/python2.7/site-packages/py/_path/common.py:411: in gen
    if p.check(dir=1) and (rec is None or rec(p))])
../../../miniconda2/envs/testenv/lib/python2.7/site-packages/_pytest/main.py:728: in _recurse
    ihook = self.gethookproxy(path)
../../../miniconda2/envs/testenv/lib/python2.7/site-packages/_pytest/main.py:632: in gethookproxy
    my_conftestmodules = pm._getconftestmodules(fspath)
../../../miniconda2/envs/testenv/lib/python2.7/site-packages/_pytest/config.py:356: in _getconftestmodules
    mod = self._importconftest(conftestpath)
../../../miniconda2/envs/testenv/lib/python2.7/site-packages/_pytest/config.py:381: in _importconftest
    raise ConftestImportFailure(conftestpath, sys.exc_info())
E   ConftestImportFailure: ImportMismatchError('sphinx_gallery.tests.conftest', '/home/travis/build/sphinx-gallery/sphinx-gallery/sphinx_gallery/tests/conftest.py', local('/home/travis/build/sphinx-gallery/sphinx-gallery/build/lib/sphinx_gallery/tests/conftest.py'))
!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!
=========================== 1 error in 4.35 seconds ============================
The command "python setup.py test" exited with 2.
```

larsoner commented:
@GaelVaroquaux:

- `sudo apt-get install texlive texlive-latex-extras latexmk` takes 1 minute
- `make latexpdf` on `tinybuild` takes 7 seconds and renders [like this](https://212-25860190-gh.circle-artifacts.com/0/tmp/circle-artifacts.NZO2VxL/Python.pdf).

Total build time is 7 minutes, so it's not a trivial addition, but still pretty quick, so it seems worth it to me.

lesteve commented:
```
=========================== 1 error in 4.35 seconds ============================
The command "python setup.py test" exited with 2.
```
Are we really running "python setup.py test" I always wondered why people do that instead of just using pytest directly.

I saw ImportMismatchError at one point in scikit-learn and I did not understand why they happened. In my case that was about testing doctests in rst files, and I worked around the problem by specifying a list of rst files to test rather than trying to use pytest automatic discovery, e.g. `pytest $(find doc -name '*.rst')` rather than `pytest doc`. Not sure whether that helps ...

larsoner commented:
I think I figured it out -- I added an exclude to `setup.cfg` that overrode the default. This made it include the `build` directory. Now that I'm excluding it, tests should pass again. (They pass locally when I reproduced the `python setup.py install && python setup.py test` pattern.)

larsoner commented:
Okay CIs are happy, ready for review/merge from my end.

larsoner commented:
I think @GaelVaroquaux gave a +1, @choldgraf also good to go from your end?

GaelVaroquaux commented:
> I think @GaelVaroquaux gave a +1,

I did!

choldgraf commented:
thx @larsoner !

QuLogic commented on sphinx_gallery/utils.py:
```diff
@@ -0,0 +1,38 @@
+from shutil import rmtree
+
+
+class _TempDir(str):
```
If this is to be used for tests, you could use the [`tmpdir` fixture](https://docs.pytest.org/en/latest/tmpdir.html#the-tmpdir-fixture).

larsoner replied:
Neat, didn't know about that

lesteve commented on doc/conf.py:
```diff
@@ -330,9 +331,6 @@ def setup(app):
     'doc_module': ('sphinx_gallery', 'numpy'),
     'reference_url': {
         'sphinx_gallery': None,
```
The `sphinx_gallery` looks a bit akward on its own, is it really needed and is there a way we can get rid of `reference_url` all together and only use intersphinx?

Titan-C replied:
Indeed, this becomes redundant with intersphinx available. I don't even think we need it. The `None` is there to tell sphinx to find the docs database locally. But we can have that done at execution time.

larsoner replied:
Ahh I see what you're saying, I can make a quick PR for that.

larsoner replied:
... but you still need to at least get `this_module` name from somewhere, otherwise I'm not sure how else you'd set the `doc_resolvers[this_module] = ...`. Without that, we don't know if we should look locally or to use `intersphinx`.

I don't see the module name in the `config` anywhere else. Storing it as a `dict` instead of a single string entry now is a bit silly, as is the name `reference_url` instead of something like `root_module` or something, but I don't mind the current version too much.

Titan-C replied:
I don't consider this very urgent. There is indeed some redundancy. And having now a dictionary for a single entry is no too elegant.
My idea is that one usually documents one local module and that module is in the "doc_module" value. Since there isn't now any need of specifying all other "reference_url" for external modules, as `intersphinx` take care of it. We could iterate the values of "doc_module" to see which modules we should try to resolve locally. 
I don't know what will happen in the case of sphinx-gallery that documents `numpy` too track the use of it in the code examples and generate the mini-galleries. I know is not related, but using the values of `doc_module` intersphinx will try to search for numpy locally and fail, and then totally independent, the value of interphinx sources in `conf.py` will take over?

weber-s commented:
Hi,

I spotted something strange in my docs. I try to use intersphinx together with SG, but it does not work unless I add   ` 'backreferences_dir': 'gen_modules/backreferences' ` to SG options. Is it the correct behaviour (sorry I'm new to sphinx world)? If so, it should be documented.

Thanks for this great job!
GwendalD

Titan-C commented:
The behavior of intersphinx should be uncorrelated with the
backreferences_dir configuration.

There is no simple way of configuring backreferences_dir, that is just a
reality we have to deal with. Please read the documentation
https://sphinx-gallery.github.io/advanced_configuration.html#adding-references-to-examples
to understand which are your options and features sphinx-gallery gives
you. In current versions of sphinx-gallery it is mandatory to choose a
value for it, we may disable it by default in the future.
